crypto/dilithium/p_dilithium_test.cc

```diff
@@ -0,0 +1,289 @@
+  EXPECT_EQ(8*(DILITHIUM3_SIGNATURE_BYTES), EVP_PKEY_bits(dilithium_pkey));
+
+  EVP_PKEY_CTX_free(dilithium_pkey_ctx);
+  EVP_PKEY_free(dilithium_pkey);
+}
+
+TEST(Dilithium3Test, Encoding) {
+  // Test Dilithium keypairs are extractable, and encode/parse correctly.
+  //generate dilithium key
+  EVP_PKEY_CTX *dilithium_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_DILITHIUM3, nullptr);
+  EVP_PKEY *dilithium_pkey = EVP_PKEY_new();
+  EVP_PKEY_keygen_init(dilithium_pkey_ctx);
+  EVP_PKEY_keygen(dilithium_pkey_ctx, &dilithium_pkey);
+  const DILITHIUM3_KEY *dilithium3Key = (DILITHIUM3_KEY *)(dilithium_pkey->pkey.ptr);
+
+  //Create a public key.
+  bssl::UniquePtr<EVP_PKEY> pubkey(EVP_PKEY_new_raw_public_key(EVP_PKEY_DILITHIUM3,
+                                                               NULL,
+                                                               dilithium3Key->pub,
+                                                               DILITHIUM3_PUBLIC_KEY_BYTES));
+  ASSERT_TRUE(pubkey);
+  EXPECT_EQ(EVP_PKEY_DILITHIUM3, EVP_PKEY_id(pubkey.get()));
+
+  // The public key must be extractable.
+  uint8_t pub_buf[1952];
+  size_t pub_len;
+  ASSERT_TRUE(EVP_PKEY_get_raw_public_key(pubkey.get(), nullptr, &pub_len));
+  EXPECT_EQ(pub_len, 1952u);
+  ASSERT_TRUE(EVP_PKEY_get_raw_public_key(pubkey.get(), pub_buf, &pub_len));
+
+  // The public key must encode properly.
+  bssl::ScopedCBB cbb;
+  uint8_t *der;
+  size_t der_len;
+  ASSERT_TRUE(CBB_init(cbb.get(), 0));
+  ASSERT_TRUE(EVP_marshal_public_key(cbb.get(), pubkey.get()));
+  ASSERT_TRUE(CBB_finish(cbb.get(), &der, &der_len));
+  bssl::UniquePtr<uint8_t> free_der(der);
+
+  // The public key must parse properly.
+  CBS cbs;
+  CBS_init(&cbs, der, der_len);
+  EVP_PKEY *dilithium_pkey_from_der = EVP_parse_public_key(&cbs);
+  ASSERT_TRUE(dilithium_pkey_from_der);
+  EXPECT_EQ(1, EVP_PKEY_cmp(dilithium_pkey, dilithium_pkey_from_der));
+
+  // Create a private key.
+  bssl::UniquePtr<EVP_PKEY> privkey(EVP_PKEY_new_raw_private_key(EVP_PKEY_DILITHIUM3,
+                                                                 NULL,
+                                                                 dilithium3Key->priv,
+                                                                 DILITHIUM3_PRIVATE_KEY_BYTES));
+  ASSERT_TRUE(privkey);
+  EXPECT_EQ(EVP_PKEY_DILITHIUM3, EVP_PKEY_id(privkey.get()));
+
+  // The private key must be extractable.
+  uint8_t priv_buf[4000];
+  size_t priv_len;
+  ASSERT_TRUE(EVP_PKEY_get_raw_private_key(privkey.get(), nullptr, &priv_len));
+  EXPECT_EQ(priv_len, 4000u);
+  ASSERT_TRUE(EVP_PKEY_get_raw_private_key(privkey.get(), priv_buf, &priv_len));
+
+  // The private key must encode properly.
+  ASSERT_TRUE(CBB_init(cbb.get(), 0));
+  ASSERT_TRUE(EVP_marshal_private_key(cbb.get(), privkey.get()));
+  ASSERT_TRUE(CBB_finish(cbb.get(), &der, &der_len));
+  free_der.reset(der);
+
+  // The private key must parse properly.
```

Comment:
Missing a comparison that dilithium_priv_from_der is equal to the original key.

Reply:
added in [2d9d0b1](https://github.com/awslabs/aws-lc/pull/674/commits/2d9d0b18a8f6c51a7b6d036e23fba0457e758e11)

Reply:
had to remove in [5e8bfc6](https://github.com/awslabs/aws-lc/pull/674/commits/5e8bfc6d4fecceb8f928c9180d92d3ef96db1e87) as `EXPECT_EQ(1, EVP_PKEY_cmp(dilithium_priv_from_der, privkey.get()));` would fail even though the DER files were exactly the same. This comparison must not be the right one.

Reply:
We'd have to [follow what they do in ed25519 ](https://github.com/awslabs/aws-lc/blob/main/crypto/evp_extra/evp_extra_test.cc#L1299-L1304). i.e. `EXPECT_EQ(Bytes(kPrivateKeyPKCS8), Bytes(der, der_len));` but this would mean hard coding in a private key/seed.

Reply:
Why does EVP_PKEY_cmp fail? Is it comparing the wrong thing? Or comparing past the end of the private key buffer?